### DISCUSSION TITLE
Open Telemetry Standard for Async Context Propagation ?

### DISCUSSION THREAD

**Discussion by rober15** [open-telemetry/community#1465] [Community Discussions]
In the micro service world the targeted communication will be via messaging brokers like kafka, rabbitmq, jms, cloud native Pub Subs

Now i wanted to know what standard exists to implement context propagation over message brokers.

- Will we go with same way as already existing for http request by using: traceparent ?
https://www.w3.org/TR/trace-context/
- Will we go with same way as defined for baggage ?
https://opentelemetry.io/docs/concepts/signals/baggage/

I recommend to go the existing way as many message brokers as well provide REST APIs.
REST API => Message Broker (publish to queue/topic)
Message Broker (publish to queue/topic) => Push Subscription to forward to REST APIs.

So if we support the already existing contexts it can easily be mapped from broker message property to http headers.

**Comment by trask** [comment]
hi @rober15!

check out https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/messaging.md#context-propagation, in particular:

> This document does not specify the exact mechanisms on how the creation context is attached/extracted to/from messages. Future versions of these conventions will give clear recommendations, following industry standards including, but not limited to [Trace Context: AMQP protocol](https://w3c.github.io/trace-context-amqp/) and [Trace Context: MQTT protocol](https://w3c.github.io/trace-context-mqtt/) once those standards reach a stable state.

@pyohannes @lmolkova @joaopgrassi is there a tracking issue in the spec repo for this that @rober15 can follow/comment on?

**Comment by pyohannes** [comment]
Thanks @trask for giving all the right pointers. [#2516](https://github.com/open-telemetry/opentelemetry-specification/issues/2516) is the issue suited best for tracking and adding requirements and thoughts.

As mentioned, there are proposals to standardize context propagation for popular messaging protocols such as [AMQP](https://w3c.github.io/trace-context-amqp/) and [MQTT](https://w3c.github.io/trace-context-mqtt/), those proposals are part of the W3C project.

> I recommend to go the existing way as many message brokers as well provide REST APIs.
> REST API => Message Broker (publish to queue/topic)
> Message Broker (publish to queue/topic) => Push Subscription to forward to REST APIs.
>
> So if we support the already existing contexts it can easily be mapped from broker message property to http headers.

This approach is supported by e. g. CloudEvents (see [here](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/cloudevents.md#conventions). This will work for messages sent and received over HTTP, but is not generally applicable to messaging.
